packages/langium/src/utils/caching.ts

```diff
@@ -0,0 +1,155 @@
+/******************************************************************************
+ * Copyright 2023 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import type { Disposable } from 'vscode-languageserver';
+import type { URI } from 'vscode-uri';
+import type { LangiumSharedServices } from '../services';
+import type { DocumentBuilder } from '../workspace';
+
+export interface CachingService {
+    createDocumentCache<K, V>(): DocumentCache<K, V>;
+    createWorkspaceCache<K, V>(): WorkspaceCache<K, V>;
+}
+
+export interface DocumentCache<K, V> extends Disposable {
+    has(documentUri: URI, key: K): boolean;
+    set(documentUri: URI, key: K, value: V): void;
+    get(documentUri: URI, key: K): V | undefined;
+    delete(documentUri: URI, key: K): boolean;
+    clear(): void;
+    clear(documentUri: URI): void;
+}
+
+export interface WorkspaceCache<K, V> extends Disposable {
+    has(key: K): boolean;
+    set(key: K, value: V): void;
+    get(key: K): V | undefined;
```

Comment:
The most obvious pattern of a cache is
```typescript
if (cache.has(key)) {
    return cache.get(key);
} else {
    const value = computeValue();
    cache.set(key, value);
    return value;
}
```
IMO every cache implementation should support this pattern, e.g. like this:
```suggestion
    get(key: K, valueProvider: () => V): V;
    get(key: K): V | undefined;
```
or move this to another method name like `access`.